src/irmin-pack/inode.ml

```diff
@@ -402,123 +482,101 @@ struct
           let entries = List.rev entries in
           Bin.Tree { depth = t.depth; length = t.length; entries }
 
-    let to_bin t =
-      let v = to_bin_v t.v in
+    let to_bin layout t =
+      let v = to_bin_v layout t.v in
       Bin.v ~stable:t.stable ~hash:t.hash v
 
     let hash t = Lazy.force t.hash
 
-    let stabilize ~find t =
+    let stabilize layout t =
       if t.stable then t
       else
         let n = length t in
         if n > Conf.stable_hash then t
         else
           let hash =
             lazy
-              (let vs = list ~find t in
+              (let vs = list layout t in
                Node.hash (Node.v vs))
           in
           { hash; stable = true; v = t.v }
 
     let hash_key = Irmin.Type.(unstage (short_hash step_t))
     let index ~depth k = abs (hash_key ~seed:depth k) mod Conf.entries
-    let entry ?target target_hash = Some { target; target_hash }
 
-    let of_bin t =
+    (** This function shouldn't be called with the [Total] layout. Unfortunately
+        the types don't enforce that. *)
```

Comment:
It's fairly simple to enforce this property by using polymorphic variants in the GADT param; that doesn't mean it's necessarily worth doing though :-)

Reply:
Yes, good idea. I've added your idea to the docstring.